src/python/pants/engine/internals/target_adaptor.py

```diff
@@ -104,7 +104,12 @@ def __init__(
     ) -> None:
         self.type_alias = type_alias
         self.name = name
-        self.kwargs = FrozenDict.deep_freeze(kwargs)
+        try:
+            self.kwargs = FrozenDict.deep_freeze(kwargs)
+        except TypeError as e:
+            # FrozenDict's ctor eagerly computes its hash. If some kwarg is unhashable it will
+            # raise a TypeError, which we enhance with the description of origin.
+            raise TypeError(f"In {__description_of_origin__}: {e}")
```

Comment:
Note: The traceback is made clearer in cases like these using `raise ... from e` to indicate that it is the same/related error re-raised.
https://docs.python.org/3/tutorial/errors.html#exception-chaining